pkg/reconciler/namespacecontroller/state.go

```diff
@@ -0,0 +1,110 @@
+// limitations under the License.
+
+package namespacecontroller
+
+import (
+	"sync"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/labels"
+)
+
+// State stores the state of the Namespace Controller.
+type State struct {
+	// eventMux is a RMMutex to protect read/write to 'isSyncingPending'.
+	// It is RMMutex because the cache is read more often than writen.
+	eventMux sync.RWMutex
+
+	// isSyncPending indicates whether the namespace controller fires up an event that requires a new sync.
+	isSyncPending bool
+
+	// selectorMux is the mutex to protect read/write to 'nsSelectors' and 'selectedNamespaces'.
+	selectorMux sync.Mutex
+
+	// nsSelectors is a map of NamespaceSelector names to the LabelSelectors.
+	nsSelectors map[string]labels.Selector
+
+	// selectedNamespaces records the dynamic/on-cluster Namespaces selected by NamespaceSelectors
+	selectedNamespaces map[string]labels.Selector
+}
+
+// NewState instantiates the namespace controller state.
+func NewState() *State {
+	return &State{
+		nsSelectors:        make(map[string]labels.Selector),
+		selectedNamespaces: make(map[string]labels.Selector),
+	}
+}
+
+// setSyncPending indicates an event is detected to trigger a new sync.
+func (s *State) setSyncPending() {
+	s.eventMux.Lock()
+	defer s.eventMux.Unlock()
+	s.isSyncPending = true
+}
+
+// IsSyncPending returns whether the state triggers a sync.
+func (s *State) IsSyncPending() bool {
+	s.eventMux.Lock()
+	defer s.eventMux.Unlock()
+	return s.isSyncPending
+}
+
+// SetSelectorCache is invoked by the reconciler to set the cache after parsing the NamespaceSelectors.
+func (s *State) SetSelectorCache(nsSelectors, selectedNamespaces map[string]labels.Selector) {
+	s.selectorMux.Lock()
+	defer s.selectorMux.Unlock()
+
+	nssMapCopy := make(map[string]labels.Selector, len(nsSelectors))
+	selectedNSMapCopy := make(map[string]labels.Selector, len(selectedNamespaces))
+	for nss, selector := range nsSelectors {
+		nssMapCopy[nss] = selector
+	}
+	for ns, selector := range selectedNamespaces {
+		selectedNSMapCopy[ns] = selector
+	}
+
+	s.nsSelectors = nssMapCopy
+	s.selectedNamespaces = selectedNSMapCopy
+}
+
+// isSelectedNamespace returns whether the namespace is previously selected.
+func (s *State) isSelectedNamespace(ns string) bool {
+	s.selectorMux.Lock()
+	defer s.selectorMux.Unlock()
+
+	_, found := s.selectedNamespaces[ns]
+	return found
+}
+
+// matchChanged returns whether the match of Namespace's labels and the cached selector has changed.
+func (s *State) matchChanged(ns *corev1.Namespace) bool {
+	s.selectorMux.Lock()
+	defer s.selectorMux.Unlock()
+
+	selector, found := s.selectedNamespaces[ns.Name]
+	// The Namespace is previously selected, but not selected anymore.
+	if found && !selector.Matches(labels.Set(ns.GetLabels())) {
+		return true
+	}
+
+	// The Namespace is not selected before, but is now selected.
+	for _, selector = range s.nsSelectors {
```

Comment:
Does the  !found need to be a criteria for the `not selected before`?

Reply:
updated